### ISSUE TITLE
feat(pipeline): Hybrid SDF→Gaussian pipeline for fast rendering of SDF geometry

### ISSUE THREAD

**Issue by hummat** [hummat/mini-mesh#14]
## Summary

Add a two-stage pipeline that trains an SDF model for high-quality geometry, then initializes 3D Gaussians on the extracted surface for real-time rendering and optional material decomposition.

**Context**: The [NeRF-3DGS evolution survey](https://arxiv.org/abs/2407.00782) identifies convergence toward hybrid approaches that combine SDF geometry quality with GS rendering speed. Mini-mesh already has both halves (SDFStudio + Nerfstudio/splatfacto) but no bridge between them.

---

## Motivation

Current mini-mesh pipeline forces a choice:
- **SDF path** → Best geometry (Neuralangelo, NeuS-facto), but slow training (hours) and no real-time rendering
- **Gaussian path** → Real-time rendering, fast training, but poor geometry (no mesh, just splats)

A hybrid pipeline gets both: SDF-quality meshes + GS-speed rendering. This is exactly what methods like NeuSG, GSDF, and HybridNeRF do internally, but we can achieve it at the pipeline level by chaining existing tools.

### Use cases
1. **Quality mesh + real-time viewer**: Train NeuS-facto → extract mesh → initialize Gaussians on mesh surface → fine-tune splatfacto for real-time viewing
2. **Material decomposition**: SDF geometry provides clean normals → initialize GaussianShader/GS-IR on mesh → decompose materials with known geometry
3. **Faster convergence**: SDF-initialized Gaussians converge faster than random/SfM init because they start on the surface

---

### Pipeline extension

```
[Existing]                              [New]
SDF train → mesh extraction ──────────→ Gaussian init on mesh → GS fine-tune → splat/mesh export
                              ↓
                         (mesh.ply)
```

**New stage between EXPORT and a second TRAIN:**

#### Step 1: Surface sampling
Sample points on the extracted SDF mesh to create a dense point cloud for Gaussian initialization:
```bash
# Sample N points on mesh surface with normals
python -c "
import trimesh
mesh = trimesh.load('mesh.ply')
points, face_idx = trimesh.sample.sample_surface(mesh, count=100000)
normals = mesh.face_normals[face_idx]
# Write as COLMAP-compatible point cloud or transforms.json init
"
```

#### Step 2: Gaussian initialization
Initialize splatfacto from the sampled point cloud instead of sparse SfM:
- Gaussian centers = sampled surface points
- Gaussian orientations = aligned to mesh normals (shortest axis = normal)
- Gaussian scales = based on local point density
- This replaces the default SfM sparse init

#### Step 3: Fine-tune with geometric constraints
Train splatfacto (or 2DGS/DN-Splatter) with:
- Surface-initialized Gaussians
- Optional normal consistency loss (normals already known from SDF)
- Fewer iterations needed (geometry already correct, just optimizing appearance)

### CLI interface

```bash
# Option A: Explicit two-stage
./run.sh --model neus-facto ...           # Stage 1: SDF
./run.sh --model splatfacto --init-from-mesh <path/to/mesh.ply>  # Stage 2: GS

# Option B: Single hybrid command
./run.sh --model hybrid-neus-splat ...    # Chains both automatically
```

### Integration points

**train.sh:**
- New `--init-from-mesh` flag that samples the mesh and creates a dense point cloud
- Pass to nerfstudio as initial point cloud (nerfstudio already supports custom init via `--pipeline.datamanager.dataparser.load-3D-points`)

**export.sh:**
- After GS fine-tuning: export both splat (for real-time) and mesh (via TSDF/Poisson from the refined Gaussians)

**run.sh:**
- If `--model hybrid-*`: chain SDF train → mesh extract → GS init → GS train → export

---

### NeuSG approach (arXiv:2312.00846)
- Uses 3DGS to provide multi-view consistent pseudo-depth/normal supervision for NeuS
- Gaussians constrained near SDF zero-level set
- Joint optimization, but conceptually: GS guides SDF geometry

### GSDF approach (arXiv:2403.16964)
- Dual-branch: SDF branch for geometry, GS branch for rendering
- Mutual supervision: SDF provides depth/normal priors to GS, GS provides photometric signal to SDF
- Key detail: GS densification guided by SDF gradient magnitude

### HybridNeRF approach (arXiv:2312.03160)
- Spatially-varying β(x) field: >95% of scene rendered as surfaces (sphere tracing), rest as volume
- Distance-adjusted Eikonal loss: 1/d² weighting for unbounded scenes
- Two-stage: global β → per-voxel β(x) with distillation to small MLP

### Practical simplification for mini-mesh
Rather than joint optimization (complex, requires deep framework changes), use **sequential chaining**:
1. Train SDF (existing pipeline) — get best geometry
2. Extract mesh (existing pipeline) — get clean surface
3. Sample mesh → initialize Gaussians → fine-tune GS (new) — get fast rendering

This is simpler than NeuSG/GSDF and leverages existing tools without modifying either framework.

---

## Alternatives Considered

1. **Joint SDF+GS training** (NeuSG/GSDF style): Requires deep integration between SDFStudio and Nerfstudio training loops. Rejected for now: too invasive, sequential chaining gets 80% of the benefit.

2. **GS-only surface methods** (#13): Complementary, not a replacement. GS surface methods (2DGS, SuGaR) produce good meshes fast, but SDF methods still produce the highest-quality geometry for challenging cases (thin structures, textureless regions).

3. **Export SDF to alternative renderers** (e.g., nvdiffrec for baked rendering): Partially covered by #7/#11. Doesn't give real-time GS rendering.

---

## Tasks

- [ ] Implement mesh → dense point cloud sampling script (`scripts/sample_mesh.py`)
- [ ] Add `--init-from-mesh` flag to `train.sh` for splatfacto init
- [ ] Wire point cloud init through nerfstudio's dataparser
- [ ] Add `--model hybrid-neus-splat` (or similar) to `run.sh` for chained execution
- [ ] Test convergence: SDF-init splatfacto vs SfM-init splatfacto (quality + speed)
- [ ] Add config files (`config/hybrid-neus-splat*.sh`)
- [ ] Document workflow in `docs/methods_and_models.md`
- [ ] Optional: Add normal consistency loss during GS fine-tune (using SDF-derived normals)

---

### Survey
- [NeRF-3DGS Evolution Survey](https://arxiv.org/abs/2407.00782) — Identifies hybrid convergence trend

### Method papers
- [NeuSG](https://arxiv.org/abs/2312.00846) — GS-guided neural implicit surface reconstruction
- [GSDF](https://arxiv.org/abs/2403.16964) — Dual-branch SDF+GS mutual supervision
- [HybridNeRF](https://arxiv.org/abs/2312.03160) — Adaptive surface/volume rendering

### Related papi papers
`neusg`, `gsdf`, `hybridnerf`, `3dgsr`, `surfacesplat`

---

## Related Issues

- #13 — GS-based surface reconstruction (complementary: GS-only path)
- #11 — PBR texture extraction (hybrid mesh + GS enables material decomposition)
- #7 — nvdiffrast Docker (enables GPU texture baking on SDF mesh)
- #5 — E2E validation (compare hybrid vs SDF-only vs GS-only quality)

**Comment by hummat** [comment]
## Cross-reference: Hybrid SDF+GS with BRDF

[GS + Discretized SDF for Relightable Assets](https://arxiv.org/abs/2507.15629) (ICCV 2025) is essentially this issue + PBR: it encodes discretized SDF values within each Gaussian as a geometry prior, then applies Principled BRDF with split-sum approximation. Achieves relightable output without extra memory beyond standard GS.

This is a concrete instantiation of the hybrid pipeline proposed here, with material decomposition on top.

Related: #13, #11, #16, sdfstudio [#10](https://github.com/hummat/sdfstudio/issues/10).
